File: Leetcode/cpp/easy/1784.cpp
/*
 * @lc app=leetcode id=1784 lang=cpp
 *
 * [1784] Check if Binary String Has at Most One Segment of Ones
 */


// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
#include <ranges>

class Solution1 {
public:
    bool checkOnesSegment(string s) {
        return ranges::fold_left(views::pairwise(s), 0, [](int acc, auto pair) {
            auto [x, y] = pair;
            return acc + (x != y ? 1 : 0);
        }) <= 1;
    }
};

class Solution2 {
public:
    bool checkOnesSegment(string s) {
        return s.find("01") == string::npos;
    }
};

// using Solution = Solution1;
using Solution = Solution2;
// @lc code=end

File: Leetcode/rust/easy/1784.rs
/*
 * @lc app=leetcode id=1784 lang=rust
 *
 * [1784] Check if Binary String Has at Most One Segment of Ones
 */


// @lcpr-template-start
struct Solution;
// @lcpr-template-end
// @lc code=start
struct Solution1;
impl Solution1 {
    pub fn check_ones_segment(s: String) -> bool {
        // as_bytes -> Vec<u8>, can be sliced
        s.as_bytes().windows(2).filter(|w| w[0] != w[1]).count() <= 1
    }
}

struct Solution2;
impl Solution2 {
    pub fn check_ones_segment(s: String) -> bool {
        !s.contains("01")
    }
}

impl Solution {
    pub fn check_ones_segment(s: String) -> bool {
        Solution1::check_ones_segment(s)
        // Solution2::check_ones_segment(s)
    }
}
// @lc code=end

